File: src/Collega.Application/Boards/BoardService.cs
using System.Text.Json;
using Collega.Application.Abstractions;
using Collega.Application.Exceptions;
using Collega.Domain.Auditing;
using Collega.Domain.Boards;
using Collega.Domain.Enums;
using Collega.Domain.Statuses;

namespace Collega.Application.Boards;

/// <summary>
/// Board configuration use cases (T021-T024). Site Admin may manage any organization; Org Admin only
/// their own. Listing and detail are available to any authenticated member of the organization;
/// create/update/reorder are admin-only (rule #5).
/// </summary>
public sealed class BoardService : IBoardService
{
    private readonly IBoardRepository _boardRepository;
    private readonly IStatusRepository _statusRepository;
    private readonly IOrganizationRepository _organizationRepository;
    private readonly IUnitOfWork _unitOfWork;
    private readonly IAuditEventWriter _auditEventWriter;
    private readonly ICurrentUserContext _currentUser;
    private readonly IClock _clock;

    public BoardService(
        IBoardRepository boardRepository,
        IStatusRepository statusRepository,
        IOrganizationRepository organizationRepository,
        IUnitOfWork unitOfWork,
        IAuditEventWriter auditEventWriter,
        ICurrentUserContext currentUser,
        IClock clock)
    {
        _boardRepository = boardRepository;
        _statusRepository = statusRepository;
        _organizationRepository = organizationRepository;
        _unitOfWork = unitOfWork;
        _auditEventWriter = auditEventWriter;
        _currentUser = currentUser;
        _clock = clock;
    }

    public async Task<IReadOnlyList<BoardListItem>> ListAsync(Guid organizationId, CancellationToken cancellationToken = default)
    {
        EnsureReadScope(organizationId);
        await EnsureOrganizationExistsAsync(organizationId, cancellationToken);

        var boards = await _boardRepository.ListByOrganizationAsync(organizationId, cancellationToken);
        return boards
            .Select(b => new BoardListItem(b.Id, b.OrganizationId, b.Name, b.AllowUserStatusUpdate, b.Swimlanes.Count))
            .ToList();
    }

    public async Task<CreateBoardResult> CreateAsync(Guid organizationId, CreateBoardCommand command, CancellationToken cancellationToken = default)
    {
        EnsureAdminScope(organizationId);
        await EnsureOrganizationExistsAsync(organizationId, cancellationToken);

        var statusLookup = await LoadStatusLookupAsync(organizationId, cancellationToken);
        var orderedStatusIds = ValidateAndOrderSwimlanes(command.Swimlanes, statusLookup);

        var now = _clock.UtcNow;
        var board = Board.Create(organizationId, command.Name, command.AllowUserStatusUpdate, orderedStatusIds, now, _currentUser.UserId);
        await _boardRepository.AddAsync(board, cancellationToken);
        await _unitOfWork.SaveChangesAsync(cancellationToken);

        await AuditAsync("BoardCreated", organizationId, board.Id, $"Board '{board.Name}' created.", now, new { board.Name, SwimlaneCount = board.Swimlanes.Count }, cancellationToken);

        var swimlanes = BuildSwimlaneDetails(board, statusLookup);
        return new CreateBoardResult(board.Id, board.Name, swimlanes);
    }

    public async Task<BoardDetail> GetByIdAsync(Guid boardId, CancellationToken cancellationToken = default)
    {
        var board = await _boardRepository.GetByIdAsync(boardId, cancellationToken)
            ?? throw new NotFoundAppException("Board not found.");

        EnsureReadScope(board.OrganizationId);

        var statusLookup = await LoadStatusLookupAsync(board.OrganizationId, cancellationToken);
        return ToDetail(board, statusLookup);
    }

    public async Task<BoardDetail> UpdateAsync(Guid boardId, UpdateBoardCommand command, CancellationToken cancellationToken = default)
    {
        var board = await _boardRepository.GetByIdAsync(boardId, cancellationToken)
            ?? throw new NotFoundAppException("Board not found.");

        EnsureAdminScope(board.OrganizationId);

        var statusLookup = await LoadStatusLookupAsync(board.OrganizationId, cancellationToken);
        var orderedStatusIds = ValidateAndOrderSwimlanes(command.Swimlanes, statusLookup);

        var now = _clock.UtcNow;
        board.Update(command.Name, command.AllowUserStatusUpdate, orderedStatusIds, now, _currentUser.UserId);
        await _unitOfWork.SaveChangesAsync(cancellationToken);

        await AuditAsync("BoardUpdated", board.OrganizationId, board.Id, $"Board '{board.Name}' updated.", now, new { board.Name, SwimlaneCount = board.Swimlanes.Count }, cancellationToken);

        return ToDetail(board, statusLookup);
    }

    public async Task ReorderSwimlanesAsync(Guid boardId, ReorderSwimlanesCommand command, CancellationToken cancellationToken = default)
    {
        var board = await _boardRepository.GetByIdAsync(boardId, cancellationToken)
            ?? throw new NotFoundAppException("Board not found.");

        EnsureAdminScope(board.OrganizationId);

        var orderedStatusIds = OrderSwimlaneInputs(command.Swimlanes);
        if (orderedStatusIds.Count != orderedStatusIds.Distinct().Count())
        {
            throw new ValidationAppException("swimlanes", new[] { "A board cannot list the same status twice." });
        }

        // A reorder cannot add or remove swimlanes — it must list exactly the current set (T024).
        var current = board.Swimlanes.Select(sl => sl.StatusId).ToHashSet();
        if (orderedStatusIds.Count != current.Count || !orderedStatusIds.All(current.Contains))
        {
            throw new ValidationAppException("swimlanes", new[] { "A reorder must list exactly the board's current swimlane statuses." });
        }

        var now = _clock.UtcNow;
        board.ReorderSwimlanes(orderedStatusIds, now, _currentUser.UserId);
        await _unitOfWork.SaveChangesAsync(cancellationToken);

        await AuditAsync("BoardSwimlanesReordered", board.OrganizationId, board.Id, $"Board '{board.Name}' swimlanes reordered.", now, null, cancellationToken);
    }

    /// <summary>
    /// Normalizes the requested swimlanes to a dense, ordered list of status ids and validates the
    /// 2-swimlane minimum (T022), distinctness, and that every status is an active status of the
    /// organization (subset support, T023).
    /// </summary>
    private static IReadOnlyList<Guid> ValidateAndOrderSwimlanes(IReadOnlyList<SwimlaneInput>? swimlanes, IReadOnlyDictionary<Guid, Status> statusLookup)
    {
        var orderedStatusIds = OrderSwimlaneInputs(swimlanes);

        if (orderedStatusIds.Count < Board.MinimumSwimlanes)
        {
            throw new ValidationAppException("swimlanes", new[] { $"A board must have at least {Board.MinimumSwimlanes} swimlanes." });
        }

        if (orderedStatusIds.Count != orderedStatusIds.Distinct().Count())
        {
            throw new ValidationAppException("swimlanes", new[] { "A board cannot list the same status twice." });
        }

        foreach (var statusId in orderedStatusIds)
        {
            if (!statusLookup.TryGetValue(statusId, out var status) || status.IsDeleted)
            {
                throw new ValidationAppException("swimlanes", new[] { "Every swimlane must reference an active status in this organization." });
            }
        }

        return orderedStatusIds;
    }

    private static List<Guid> OrderSwimlaneInputs(IReadOnlyList<SwimlaneInput>? swimlanes) =>
        (swimlanes ?? Array.Empty<SwimlaneInput>())
            .OrderBy(s => s.Order)
            .Select(s => s.StatusId)
            .ToList();

    private async Task<IReadOnlyDictionary<Guid, Status>> LoadStatusLookupAsync(Guid organizationId, CancellationToken cancellationToken)
    {
        // Include deleted so board detail can still surface a prior status name if one was later
        // removed (rule #8); create/update validation rejects deleted statuses separately.
        var statuses = await _statusRepository.ListByOrganizationAsync(organizationId, includeDeleted: true, cancellationToken);
        return statuses.ToDictionary(s => s.Id);
    }

    private static BoardDetail ToDetail(Board board, IReadOnlyDictionary<Guid, Status> statusLookup) =>
        new(board.Id, board.OrganizationId, board.Name, board.AllowUserStatusUpdate, BuildSwimlaneDetails(board, statusLookup));

    private static IReadOnlyList<SwimlaneDetail> BuildSwimlaneDetails(Board board, IReadOnlyDictionary<Guid, Status> statusLookup) =>
        board.Swimlanes
            .OrderBy(sl => sl.DisplayOrder)
            .Select(sl =>
            {
                var status = statusLookup.GetValueOrDefault(sl.StatusId);
                return new SwimlaneDetail(
                    sl.StatusId,
                    status?.Name ?? string.Empty,
                    status?.Color ?? string.Empty,
                    sl.DisplayOrder,
                    status?.IsDeleted ?? false);
            })
            .ToList();

    private async Task EnsureOrganizationExistsAsync(Guid organizationId, CancellationToken cancellationToken)
    {
        var organization = await _organizationRepository.GetByIdAsync(organizationId, cancellationToken)
            ?? throw new NotFoundAppException("Organization not found.");
        _ = organization;
    }

    private void EnsureReadScope(Guid organizationId)
    {
        var role = RequireAuthenticatedRole();
        if (role == Role.SiteAdmin)
        {
            return;
        }

        if (_currentUser.OrganizationId == organizationId)
        {
            return;
        }

        throw new NotFoundAppException("Board not found.");
    }

    private void EnsureAdminScope(Guid organizationId)
    {
        var role = RequireAuthenticatedRole();

        // Rule 25: a Site Admin acting as themselves may not mutate org-owned content — View As is
        // the path. No impersonation branch is needed: while a session is live this role is the
        // target's, so the guard does not fire. Read paths use EnsureReadScope and are unaffected.
        _currentUser.EnsureNotDirectSiteAdmin();

        // No SiteAdmin pass-through below, deliberately: the guard above has already refused that
        // role, so a `return` here would be dead code that reads like a live bypass. Should the
        // guard ever move, falling through to the OrgAdmin check denies rather than admits.
        if (role == Role.OrgAdmin)
        {
            if (_currentUser.OrganizationId == organizationId)
            {
                return;
            }

            throw new NotFoundAppException("Board not found.");
        }

        throw new ForbiddenAppException("You are not allowed to manage boards in this organization.");
    }

    private Role RequireAuthenticatedRole()
    {
        if (!_currentUser.IsAuthenticated || _currentUser.Role is null)
        {
            throw new UnauthorizedAppException("Caller identity could not be resolved.");
        }

        return _currentUser.Role.Value;
    }

    private async Task AuditAsync(
        string eventType,
        Guid organizationId,
        Guid boardId,
        string message,
        DateTime nowUtc,
        object? metadata,
        CancellationToken cancellationToken)
    {
        var metadataJson = metadata is null ? null : JsonSerializer.Serialize(metadata);
        // Rule 14: while acting as someone, the real administrator is the actor and the target
        // moves to OnBehalfOfUserId — an audit row must never read as though the target did it.
        var attribution = _currentUser.AttributeAudit(_currentUser.UserId);
        var auditEvent = AuditEvent.Create(eventType, "Board", message, nowUtc, organizationId, attribution.ActorUserId, boardId, metadataJson, attribution.OnBehalfOfUserId);
        await _auditEventWriter.WriteAsync(auditEvent, cancellationToken);
    }
}

File: packages/application/src/boards/board-service.ts
// Board configuration use cases (SPEC/20-feature-boards-and-statuses.md "Board Rules",
// SPEC/30-Contracts.md "Board Contracts"). Site Admin may manage any organization; Org Admin
// only their own. Listing and detail are available to any authenticated member of the
// organization; create/update/reorder are admin-only.

import { randomUUID } from 'node:crypto'
import {
  type Board,
  BoardInvariantError,
  createBoard,
  MIN_SWIMLANES,
  reorderBoardSwimlanes,
  updateBoard,
} from '@collega/domain/boards'
import { Role } from '@collega/domain/enums'
import type { Status } from '@collega/domain/statuses'
import {
  type AuditEventWriter,
  attributeAudit,
  type Clock,
  type CurrentUserContext,
  ensureNotDirectSiteAdmin,
  ForbiddenError,
  NotFoundError,
  UnauthorizedError,
  type UnitOfWork,
  ValidationError,
} from '../common/index.js'
import type { StatusRepository } from '../statuses/index.js'
import type {
  BoardDetail,
  BoardListItem,
  CreateBoardCommand,
  CreateBoardResult,
  ReorderSwimlanesCommand,
  SwimlaneDetail,
  SwimlaneInput,
  UpdateBoardCommand,
} from './models.js'
import type { BoardRepository, OrganizationExistenceLookup } from './ports.js'

export class BoardService {
  constructor(
    private readonly boards: BoardRepository,
    private readonly statuses: StatusRepository,
    private readonly organizations: OrganizationExistenceLookup,
    private readonly unitOfWork: UnitOfWork,
    private readonly auditEvents: AuditEventWriter,
    private readonly currentUser: CurrentUserContext,
    private readonly clock: Clock,
  ) {}

  async list(organizationId: string): Promise<readonly BoardListItem[]> {
    this.ensureReadScope(organizationId)
    await this.ensureOrganizationExists(organizationId)

    const boards = await this.boards.listByOrganization(organizationId)
    return [...boards].sort(compareBoardsForListing).map((board) => ({
      boardId: board.id,
      organizationId: board.organizationId,
      name: board.name,
      allowUserStatusUpdate: board.allowUserStatusUpdate,
      swimlaneCount: board.swimlanes.length,
    }))
  }

  async create(organizationId: string, command: CreateBoardCommand): Promise<CreateBoardResult> {
    this.ensureAdminScope(organizationId)
    await this.ensureOrganizationExists(organizationId)

    const statusLookup = await this.loadStatusLookup(organizationId)
    const orderedStatusIds = validateAndOrderSwimlanes(command.swimlanes, statusLookup)

    const now = this.clock.now()
    const board = runDomain(createBoard, {
      id: randomUUID(),
      organizationId,
      name: command.name,
      allowUserStatusUpdate: command.allowUserStatusUpdate,
      orderedStatusIds,
      nowUtc: now,
      actorUserId: this.currentUser.userId,
    })
    await this.boards.add(board)
    await this.unitOfWork.saveChanges()

    await this.audit(
      'BoardCreated',
      organizationId,
      board.id,
      `Board '${board.name}' created.`,
      now,
      { name: board.name, swimlaneCount: board.swimlanes.length },
    )

    return {
      boardId: board.id,
      name: board.name,
      swimlanes: buildSwimlaneDetails(board, statusLookup),
    }
  }

  async getById(boardId: string): Promise<BoardDetail> {
    const board = await this.boards.getById(boardId)
    if (board === null) {
      throw new NotFoundError('Board not found.')
    }

    this.ensureReadScope(board.organizationId)

    const statusLookup = await this.loadStatusLookup(board.organizationId)
    return toDetail(board, statusLookup)
  }

  async update(boardId: string, command: UpdateBoardCommand): Promise<BoardDetail> {
    const existing = await this.boards.getById(boardId)
    if (existing === null) {
      throw new NotFoundError('Board not found.')
    }

    this.ensureAdminScope(existing.organizationId)

    const statusLookup = await this.loadStatusLookup(existing.organizationId)
    const orderedStatusIds = validateAndOrderSwimlanes(command.swimlanes, statusLookup)

    const now = this.clock.now()
    const board = runDomain(
      updateBoard,
      existing,
      {
        name: command.name,
        allowUserStatusUpdate: command.allowUserStatusUpdate,
        orderedStatusIds,
      },
      now,
      this.currentUser.userId,
    )
    await this.boards.save(board)
    await this.unitOfWork.saveChanges()

    await this.audit(
      'BoardUpdated',
      board.organizationId,
      board.id,
      `Board '${board.name}' updated.`,
      now,
      { name: board.name, swimlaneCount: board.swimlanes.length },
    )

    return toDetail(board, statusLookup)
  }

  async reorderSwimlanes(boardId: string, command: ReorderSwimlanesCommand): Promise<void> {
    const existing = await this.boards.getById(boardId)
    if (existing === null) {
      throw new NotFoundError('Board not found.')
    }

    this.ensureAdminScope(existing.organizationId)

    const orderedStatusIds = orderSwimlaneInputs(command.swimlanes)
    if (new Set(orderedStatusIds).size !== orderedStatusIds.length) {
      throw new ValidationError('One or more fields are invalid.', {
        swimlanes: ['A board cannot list the same status twice.'],
      })
    }

    // A reorder cannot add or remove swimlanes - it must list exactly the current set.
    const current = new Set(existing.swimlanes.map((swimlane) => swimlane.statusId))
    const requested = new Set(orderedStatusIds)
    const sameSet =
      orderedStatusIds.length === current.size && [...requested].every((id) => current.has(id))
    if (!sameSet) {
      throw new ValidationError('One or more fields are invalid.', {
        swimlanes: ["A reorder must list exactly the board's current swimlane statuses."],
      })
    }

    const now = this.clock.now()
    const board = runDomain(
      reorderBoardSwimlanes,
      existing,
      orderedStatusIds,
      now,
      this.currentUser.userId,
    )
    await this.boards.save(board)
    await this.unitOfWork.saveChanges()

    await this.audit(
      'BoardSwimlanesReordered',
      board.organizationId,
      board.id,
      `Board '${board.name}' swimlanes reordered.`,
      now,
      null,
    )
  }

  private async loadStatusLookup(organizationId: string): Promise<ReadonlyMap<string, Status>> {
    // Include deleted so board detail can still surface a prior status name if one was later
    // removed (rule #8); create/update validation rejects deleted statuses separately.
    const statuses = await this.statuses.listByOrganization(organizationId, true)
    return new Map(statuses.map((status) => [status.id, status]))
  }

  private async ensureOrganizationExists(organizationId: string): Promise<void> {
    if (!(await this.organizations.existsById(organizationId))) {
      throw new NotFoundError('Organization not found.')
    }
  }

  private ensureReadScope(organizationId: string): void {
    const role = this.requireAuthenticatedRole()
    if (role === Role.SiteAdmin) {
      return
    }
    if (this.currentUser.organizationId === organizationId) {
      return
    }
    throw new NotFoundError('Board not found.')
  }

  private ensureAdminScope(organizationId: string): void {
    const role = this.requireAuthenticatedRole()

    // Rule 25: a Site Admin acting as themselves may not mutate org-owned content - View As is
    // the path. No impersonation branch is needed: while a session is live this role is the
    // target's, so the guard does not fire. Read paths use ensureReadScope and are unaffected.
    ensureNotDirectSiteAdmin(this.currentUser)

    // No SiteAdmin pass-through below, deliberately: the guard above has already refused that
    // role, so a `return` here would be dead code that reads like a live bypass.
    if (role === Role.OrgAdmin) {
      if (this.currentUser.organizationId === organizationId) {
        return
      }
      throw new NotFoundError('Board not found.')
    }

    throw new ForbiddenError('You are not allowed to manage boards in this organization.')
  }

  private requireAuthenticatedRole(): Role {
    if (!this.currentUser.isAuthenticated || this.currentUser.role === null) {
      throw new UnauthorizedError('Caller identity could not be resolved.')
    }
    return this.currentUser.role
  }

  private async audit(
    eventType: string,
    organizationId: string,
    boardId: string,
    message: string,
    occurredAtUtc: Date,
    metadata: Record<string, unknown> | null,
  ): Promise<void> {
    // Rule 14: while acting as someone, the real administrator is the actor and the target moves
    // to onBehalfOfUserId - an audit row must never read as though the target did it.
    const attribution = attributeAudit(this.currentUser, this.currentUser.userId)
    await this.auditEvents.write({
      eventType,
      entityType: 'Board',
      message,
      occurredAtUtc,
      organizationId,
      attribution,
      entityId: boardId,
      metadataJson: metadata === null ? null : JSON.stringify(metadata),
    })
  }
}

function orderSwimlaneInputs(swimlanes: readonly SwimlaneInput[]): readonly string[] {
  return [...swimlanes].sort((a, b) => a.order - b.order).map((swimlane) => swimlane.statusId)
}

/**
 * Normalizes the requested swimlanes to a dense, ordered list of status ids and validates the
 * 2-swimlane minimum, distinctness, and that every status is an active status of the
 * organization (subset support).
 */
function validateAndOrderSwimlanes(
  swimlanes: readonly SwimlaneInput[],
  statusLookup: ReadonlyMap<string, Status>,
): readonly string[] {
  const orderedStatusIds = orderSwimlaneInputs(swimlanes)

  if (orderedStatusIds.length < MIN_SWIMLANES) {
    throw new ValidationError('One or more fields are invalid.', {
      swimlanes: [`A board must have at least ${MIN_SWIMLANES} swimlanes.`],
    })
  }

  if (new Set(orderedStatusIds).size !== orderedStatusIds.length) {
    throw new ValidationError('One or more fields are invalid.', {
      swimlanes: ['A board cannot list the same status twice.'],
    })
  }

  for (const statusId of orderedStatusIds) {
    const status = statusLookup.get(statusId)
    if (status === undefined || status.isDeleted) {
      throw new ValidationError('One or more fields are invalid.', {
        swimlanes: ['Every swimlane must reference an active status in this organization.'],
      })
    }
  }

  return orderedStatusIds
}

/**
 * Wraps a domain transition so a `BoardInvariantError` (a plain-Error invariant violation -
 * packages/domain imports nothing, so it cannot throw the kernel's `ValidationError` itself)
 * surfaces as a proper field-level 400, mirroring how .NET let `Board`'s `ArgumentException`
 * reach the API only through a `ValidationAppException` raised in the service, never directly
 * (SPEC/decisions.md 2026-09-06 "Wave B conventions"; B3's `runDomain` is the reference).
 *
 * Takes `fn` and its arguments separately, rather than a `() => T` thunk, so a narrowed `let`
 * or `const` binding is read as a plain argument expression instead of inside a closure body.
 */
function runDomain<Args extends readonly unknown[], T>(fn: (...args: Args) => T, ...args: Args): T {
  try {
    return fn(...args)
  } catch (error) {
    if (error instanceof BoardInvariantError) {
      throw new ValidationError('One or more fields are invalid.', {
        [error.field]: [error.message],
      })
    }
    throw error
  }
}

function toDetail(board: Board, statusLookup: ReadonlyMap<string, Status>): BoardDetail {
  return {
    boardId: board.id,
    organizationId: board.organizationId,
    name: board.name,
    allowUserStatusUpdate: board.allowUserStatusUpdate,
    swimlanes: buildSwimlaneDetails(board, statusLookup),
  }
}

function buildSwimlaneDetails(
  board: Board,
  statusLookup: ReadonlyMap<string, Status>,
): readonly SwimlaneDetail[] {
  return [...board.swimlanes]
    .sort((a, b) => a.displayOrder - b.displayOrder)
    .map((swimlane) => {
      const status = statusLookup.get(swimlane.statusId)
      return {
        statusId: swimlane.statusId,
        statusName: status?.name ?? '',
        statusColor: status?.color ?? '',
        order: swimlane.displayOrder,
        statusIsDeleted: status?.isDeleted ?? false,
      }
    })
}

/**
 * Total order for the board list: name, then creation time - never id. Nothing enforces unique
 * board names, so a name tie is possible; the golden capture's collision note found endpoints
 * elsewhere breaking such ties on generated id, which disagrees between two databases seeded
 * from the same data (SPEC/50-typescript-migration.md). Creation time is stable and meaningful
 * even though the .NET list only ordered by name.
 */
function compareBoardsForListing(a: Board, b: Board): number {
  const byName = compareStrings(a.name, b.name)
  if (byName !== 0) {
    return byName
  }
  return a.createdAtUtc.getTime() - b.createdAtUtc.getTime()
}

function compareStrings(a: string, b: string): number {
  if (a < b) {
    return -1
  }
  if (a > b) {
    return 1
  }
  return 0
}
